### PR TITLE
Issue #3 added car's fuel consumption units

### PR DESCRIPTION
**Pull request by uttpal**
Added measurement units
g/km, kg/km, g/mi, kg/mi, ltr/km, gal/km
for users to choose from.

### CODE REVIEW

File: Source/options.html [hunk 1]
```diff
@@ -26,6 +26,14 @@
             To improve the accuracy of the estimation, <br />
             please fill in below how much CO<sub>2</sub> your car emits (in g/km): <br />
```
Comment (Ceilican) [inline Source/options.html]: this line needs to be changed too. I suggest:

"... please fill in below how much fuel your car consumes or how much CO<sub>2</sub> it emits: "

**Comment by Ceilican** [comment]
Thanks, Uttpal! That's a great start!  Could you just do the change I suggested in my line comment? Then I will merge your pull request.

File: Source/options.js [hunk 2]
```diff
@@ -1,28 +1,49 @@
...
+            carbonEmission = 0.621371 * carbonEmission;
+            break;
+        case "kg/mi":
+            carbonEmission = 621.371 * carbonEmission;
+            break;
+        case "ltr/km":
+            carbonEmission = 2320 * carbonEmission;
+            break;
+        case "gal/km":
+            carbonEmission = 8782.155 * carbonEmission;
```
Comment (Ceilican) [inline Source/options.js]: Where did you obtain these conversion factors? On which assumptions do they depend? Could you please add some comments in the code, indicating the source that you used?

Thanks!

**Comment by Ceilican** [comment]
Also, some people prefer to specify not consumption/emission, but its inverse: fuel efficiency. In other words, they prefer to say how many kilometers per liter the car can go, instead of saying how many liters per kilometer the car needs. How would you propose to deal with this, from a user interface point of view? How would you allow the user to specify fuel efficiency?

**Comment by Ceilican** [comment]
It is important that you introduce yourself in our mailing-list: https://groups.google.com/forum/#!forum/cse-at-tu-wien-gsoc-2016

**Comment by Ceilican** [comment]
Sorry for the broken link. I will post the code below:

```

var fuels = {
  diesel : 0,
  gasoline : 1,
  lpg : 2,
  e10 : 3,
  e25 : 4,
  e85 : 5,
  ethanol : 6,
  biodiesel : 7
};

// Static Info
// ===========
//
// Source: http://www.biomassenergycentre.org.uk/portal/page?_pageid=75,163182&_dad=portal&_schema=PORTAL


// Energy Density, in MJ/l
var energyDensity = {}; 
energyDensity[fuels.diesel] = 36;
energyDensity[fuels.gasoline] = 32;
energyDensity[fuels.lpg] = 24;
energyDensity[fuels.e10] = (0.10 * 21) + (0.90 * 32) ;
energyDensity[fuels.e25] = (0.25 * 21) + (0.75 * 32);
energyDensity[fuels.e85] = (0.85 * 21) + (0.15 * 32);
energyDensity[fuels.ethanol] = 21;
energyDensity[fuels.biodiesel] = 33;

// CO2 emission on combustion, in g/l
var CO2Emission = {}; 
CO2Emission[fuels.diesel] = 2614;
CO2Emission[fuels.gasoline] = 2328;
CO2Emission[fuels.lpg] = 1533;
CO2Emission[fuels.e10] = (0.10 * 1503) + (0.90 * 2328) ;
CO2Emission[fuels.e25] = (0.25 * 1503) + (0.75 * 2328);
CO2Emission[fuels.e85] = (0.85 * 1503) + (0.15 * 2328);
CO2Emission[fuels.ethanol] = 1503;
CO2Emission[fuels.biodiesel] = 2486;

// approximate life-cycle GHG emission, in CO2 eq. g/l
var GHGEmission = {}; 
GHGEmission[fuels.diesel] = 3128;
GHGEmission[fuels.gasoline] = 2600;
GHGEmission[fuels.lpg] = 1533 * (2600/2328);    //assuming it grows as for gasoline
GHGEmission[fuels.e10] = (0.10 * (724 + 511)/2) + (0.90 * 2328) ;
GHGEmission[fuels.e25] = (0.25 * (724 + 511)/2) + (0.75 * 2328);
GHGEmission[fuels.e85] = (0.85 * (724 + 511)/2) + (0.15 * 2328);
GHGEmission[fuels.ethanol] = (724 + 511)/2;     // wheat = 511, sugar beet = 724 
GHGEmission[fuels.biodiesel] = (1334 + 437/2);  // waste vegetable oil = 437 , rapeseed oil = 1334


// Initialization
// ==============

// State variables
var fuel = 1;
var fuelEfficiency = 9;
var isUnitSystemImperial = false;


function initialize(f, fE, uS) {
  fuel = f;
  fuelEfficiency = fE;
  isUnitSystemImperial = uS;
  initializeMap();
  updateUI();
}

var directionsDisplay = new google.maps.DirectionsRenderer({draggable: true});
var directionsService = new google.maps.DirectionsService();


function initializeMap() {
  var worldCenter = new google.maps.LatLng(0, 0);
  var mapOptions = {
    zoom:2,
    mapTypeId: google.maps.MapTypeId.ROADMAP,
    center: worldCenter
  }
  var map = new google.maps.Map(document.getElementById('map_canvas'), mapOptions);
  directionsDisplay.setMap(map);
	
  google.maps.event.addListener(directionsDisplay, 'directions_changed', function() {
   updateUI();
  });
}



// Functions called by buttons and events of the UI
// ================================================

function calcRoute() {
  var start = document.getElementById('start').value;
  var end = document.getElementById('end').value;
  var request = {
	origin:start,
	destination:end,
	travelMode: google.maps.DirectionsTravelMode.DRIVING
  };
  directionsService.route(request, function(response, status) {
    if (status == google.maps.DirectionsStatus.OK) {
	  directionsDisplay.setDirections(response);
	  updateUI();
    }
  });
}


function keyPress(event) {
  if (event.keyCode == 13) calcRoute();
} 


function offset() {
  var emissionGHG = computeEmissions(getDistance(directionsDisplay.directions.routes[0])).GHG;
  var trees = Math.round(emissionGHG/8200); // Trees necessary for absorbing the emissions in one year
  
  window.location.href = "http://maps.oiworld.org/plant?trees=" + trees;
}
  
  
function changeFuelEfficiency() {
  fuelEfficiency = document.getElementById('fuel_efficiency_input').value; 
  updateUI();
}  
  
  
function changeUnitSystem() {
  isUnitSystemImperial = !isUnitSystemImperial;
  if (isUnitSystemImperial) fuelEfficiency = Math.round(fuelEfficiency * 2352.15)/1000;
  else fuelEfficiency = Math.round((fuelEfficiency * 1000) / 2.35215)/1000;   
  updateUI();
} 


function changeFuel() {
  var previousFuel = fuel;
  
  if (fuel < (fuelNames.length - 1)) fuel = fuel + 1;
  else fuel = 0;
  
  fuelEfficiency = fuelEfficiency * (energyDensity[fuel] / energyDensity[previousFuel]);
  
  updateUI();
} 
   


// Update of the UI
// ================  
  
function updateUI() {
  document.getElementById("unit_system_button").innerHTML = unitSystemToString();
  document.getElementById("fuel_type_button").innerHTML = fuelNames[fuel];
  document.getElementById('fuel_efficiency_input').value = Math.round(fuelEfficiency*10)/10;

  if (directionsDisplay.directions) {
    var distance = getDistance(directionsDisplay.directions.routes[0]);
    document.getElementById("distance").innerHTML = distanceToString(distance);
	
    var duration = getDuration(directionsDisplay.directions.routes[0]);
    document.getElementById("duration").innerHTML = durationToString(duration);
	
    var emissions = computeEmissions(distance);	
    document.getElementById("CO2").innerHTML = emissionToString(emissions.CO2, "CO<sub>2</sub>");
    document.getElementById("GHG").innerHTML = emissionToString(emissions.GHG, "CO<sub>2</sub>e <sub>(lifecycle)</sub>");
  
    document.getElementById("offset_button").style.visibility = "visible";
  }
}


function unitSystemToString() {
  return isUnitSystemImperial ? "mpg" : "km/l";
}   
  
   
function emissionToString(amount, poluent) { // amount in grams
  if (isUnitSystemImperial) {
    if (amount < 50000) return Math.round(amount*2.2046/100)/10 + " lb of " + poluent;
    else return Math.round(amount*2.2046/1000) + " lb of " + poluent;
  }
  else {
    if (amount < 1000) return amount + " g of " + poluent;
    else if (amount < 50000) return Math.round(amount/100)/10 + " kg of " + poluent;
    else return Math.round(amount/1000) + " kg of " + poluent;
  }
}


function distanceToString(distance) {
  if (isUnitSystemImperial) {
    if (distance < 500) return Math.round(distance/0.9144) + " yd";
    else if (distance < 50000) return Math.round(distance/160.93)/10 + " mi";
    else return Math.round(distance/1609.3) + " mi";
  }
  else {
    if (distance < 1000) return distance + " m";
    else if (distance < 50000) return Math.round(distance/100)/10 + " km";
    else return Math.round(distance/1000) + " km";
  }
}
 
    
function durationToString(duration) {
  var hours = Math.floor(duration / 3600);
  var minutes = Math.round((duration - (hours * 3600))/60); 
  return ((hours == 0) ? "" : (hours + " h ")) + minutes + " min";
}  
  
  

// Functions to calculate data from the route
// ==========================================
  
function getDistance(route) {
  var distance = 0;
  for (i = 0; i < route.legs.length; i++) {
    distance += route.legs[i].distance.value;
  }
  return distance; // in meters
}


function getDuration(route) {
  var duration = 0;
  for (i = 0; i < route.legs.length; i++) {
    duration += route.legs[i].duration.value;
  }
  return duration; // in seconds
}


function computeFuelAmount(distance) {
  var efficiency = isUnitSystemImperial ? (fuelEfficiency / 2.35215) : fuelEfficiency; // in km/l
  return ((distance/1000)/efficiency); // in liters
}


function computeEmissions(distance) {
  var fuelAmount = computeFuelAmount(distance);
  return {
    CO2: fuelAmount * CO2Emission[fuel],
    GHG: fuelAmount * GHGEmission[fuel]
  }
}
  

```

**Comment by Ceilican** [comment]
Have a look on my comments here as well: https://github.com/OiWorld/CarbonFootprintGoogleMaps/issues/3#issuecomment-190880428

**Comment by uttpal** [comment]
Added way for user to use car's fuel consumption units in litres per km, or kilometres per litre or miles per gallon and others

**Comment by PrateekGupta1509** [comment]
I dealt with this in a slightly different manner which is a little more user friendly.  Providing 3 tabs for each possible category of input: Fuel consumption, Fuel efficiency, CO2 Emission; taking into the account different fuel types and units .
Please have a look at my pull request https://github.com/OiWorld/CarbonFootprintGoogleMaps/pull/14
Suggest me if there is any changes or problem in the request.

**Comment by Ceilican** [comment]
@uttpal @PrateekGupta1509  I had a quick look on both of your proposals, but I will need to look into them more deeply tomorrow, before deciding what to do. So, please be patient. I will probably look into combining the best ideas from both of you. So, if you have ideas on how to do that, please share them here.

In the meanwhile, try not to work on the same issue at the same time. And thank you very much for your contributions!

**Comment by Ceilican** [comment]
@uttpal @PrateekGupta1509 I had time to analyse your pull requests more carefully today. Both successfully address issue #9 , but @PrateekGupta1509 's pull request also addresses issue #3 , which also asks for allowing the user to specify the consumption, besides emission.

Regarding issue #9, both pull requests have minor UI issues. For instance, when the user changes the units, the value is not converted.

Compared to the example code that I provided in issue https://github.com/OiWorld/CarbonFootprintGoogleMaps/issues/3, the code from pull request https://github.com/OiWorld/CarbonFootprintGoogleMaps/pull/14 seems more convoluted and verbose than it needs to be.

I'm inclined to merge pull request https://github.com/OiWorld/CarbonFootprintGoogleMaps/pull/14/files because it has more features than pull request https://github.com/OiWorld/CarbonFootprintGoogleMaps/pull/13, but only after these minor issues have been fixed.

In the meanwhile, pay attention to this new puzzle too: https://github.com/OiWorld/CarbonFootprintGoogleMaps/issues/16 .

**Comment by PrateekGupta1509** [comment]
@Ceilican: I would like to know if you want change in value on change of units in case of CO2 emission or for fuel  consumption and fuel efficiency as well ?

**Comment by Ceilican** [comment]
@PrateekGupta: In all cases. I think the code that I shared in the comments had a function that did conversions and then updated the UI. Have a look there.

**Comment by PrateekGupta1509** [comment]
@Ceilican if we allow automatic conversion on change of unit in select box. It will change the value that user entered in input  before entering the units. So he must first select his unit and then enter the value.

**Comment by Ceilican** [comment]
I'm closing this pull request, because it s functionality was superseded by the functionality of pull request
https://github.com/OiWorld/CarbonFootprintGoogleMaps/pull/30 .

**Pull request closed without merging**